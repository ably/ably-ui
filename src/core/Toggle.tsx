import * as Switch from "@radix-ui/react-switch";
import React from "react";
import cn from "./utils/cn";

type ToggleProps = {
  id: string;
  size?: "sm" | "lg";
  label?: string;
  className?: string;
  onCheckedChange?: (checked: boolean) => void;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

const Toggle: React.FC<ToggleProps> = ({
  id,
  size = "lg",
  label,
  className,
  ...props
}) => {
  const rootSize = size === "sm" ? "w-[42px] h-24" : "w-[56px] h-32";
  const thumbSize =
    size === "sm"
      ? "w-[21px] h-[21px] translate-x-1 data-[state=checked]:translate-x-[20px]"
      : "h-[28px] w-[28px] translate-x-2 data-[state=checked]:translate-x-[26px]";

  return (
    <div className="flex items-center">
      <Switch.Root
        className={cn(
          "p-0 bg-neutral-600 rounded-full relative inline-block transition-colors data-[disabled]:bg-gui-unavailable data-[disabled]:cursor-not-allowed data-[state=checked]:bg-orange-600 focus-base",

          className,
          rootSize,
        )}
        id={id}
        {...props}
      >
        <Switch.Thumb
          className={cn(
            `block bg-white data-[disabled]:bg-neutral-500 rounded-full drop-shadow-toggle transition-transform`,
            thumbSize,
          )}
        />
      </Switch.Root>
      {label ? (
        <label className="ml-16" htmlFor={id}>
          {label}
        </label>
      ) : null}
    </div>
  );
};

export default Toggle;
