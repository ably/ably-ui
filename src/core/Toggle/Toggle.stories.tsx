import React from "react";
import { Meta, StoryFn } from "@storybook/react-vite";
import Toggle from "../Toggle";

export default {
  title: "Components/Toggle",
  component: Toggle,
  tags: ["autodocs"],
} as Meta;

const Template: StoryFn<typeof Toggle> = (args) => <Toggle {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const Checked = Template.bind({});
Checked.args = {
  defaultChecked: true,
  disabled: false,
};

export const Disabled = Template.bind({});
Disabled.args = {
  defaultChecked: false,
  disabled: true,
};
