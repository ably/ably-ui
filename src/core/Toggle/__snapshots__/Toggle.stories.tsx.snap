// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Toggle Checked smoke-test 1`] = `
<div class="flex items-center">
  <button type="button"
          role="switch"
          aria-checked="true"
          data-state="checked"
          value="on"
          class="p-0 h-32 w-[56px] bg-neutral-600 rounded-full relative inline-block transition-colors data-[disabled]:bg-gui-unavailable data-[disabled]:cursor-not-allowed data-[state=checked]:bg-orange-600 focus:outline-none focus-visible:outline-offset-0 focus-visible:outline-4 focus-visible:outline-gui-focus"
  >
    <span data-state="checked"
          class="block h-[28px] w-[28px] bg-white data-[disabled]:bg-neutral-500 rounded-full drop-shadow-toggle translate-x-2 data-[state=checked]:translate-x-[26px] transition-transform"
    >
    </span>
  </button>
</div>
`;

exports[`Components/Toggle Default smoke-test 1`] = `
<div class="flex items-center">
  <button type="button"
          role="switch"
          aria-checked="false"
          data-state="unchecked"
          value="on"
          class="p-0 h-32 w-[56px] bg-neutral-600 rounded-full relative inline-block transition-colors data-[disabled]:bg-gui-unavailable data-[disabled]:cursor-not-allowed data-[state=checked]:bg-orange-600 focus:outline-none focus-visible:outline-offset-0 focus-visible:outline-4 focus-visible:outline-gui-focus"
  >
    <span data-state="unchecked"
          class="block h-[28px] w-[28px] bg-white data-[disabled]:bg-neutral-500 rounded-full drop-shadow-toggle translate-x-2 data-[state=checked]:translate-x-[26px] transition-transform"
    >
    </span>
  </button>
</div>
`;

exports[`Components/Toggle Disabled smoke-test 1`] = `
<div class="flex items-center">
  <button type="button"
          role="switch"
          aria-checked="false"
          data-state="unchecked"
          data-disabled
          disabled
          value="on"
          class="p-0 h-32 w-[56px] bg-neutral-600 rounded-full relative inline-block transition-colors data-[disabled]:bg-gui-unavailable data-[disabled]:cursor-not-allowed data-[state=checked]:bg-orange-600 focus:outline-none focus-visible:outline-offset-0 focus-visible:outline-4 focus-visible:outline-gui-focus"
  >
    <span data-state="unchecked"
          data-disabled
          class="block h-[28px] w-[28px] bg-white data-[disabled]:bg-neutral-500 rounded-full drop-shadow-toggle translate-x-2 data-[state=checked]:translate-x-[26px] transition-transform"
    >
    </span>
  </button>
</div>
`;
